PR for SRVLOGIC-98: Add "Authentication for OpenAPI services" section in the OSL docs (#92245)
openshift/openshift-docs: modules/serverless-logic-security-example-oauth-authentication.adoc

                  "type": "string"
                }
              }
            }
          }
        },
        "security": [
          {
            "oauth-example": []
          }
        ]
      }
    }
  },
  "components": {
    "securitySchemes": {
      "oauth-example": {
        "type": "oauth2",
        "flows": {
          "clientCredentials": {
            "authorizationUrl": "https://example.com/oauth",
            "tokenUrl": "https://example.com/oauth/token",
            "scopes": {}
          }
        }
      }
    }
  }
}
----

In this example, the `sayHelloOauth2` operation is protected by the `oauth-example` security scheme, which uses the `clientCredentials` flow for token-based authentication.

[id="serverless-logic-security-oauth-support-oidc-client-filter-extention_{context}"]
== OAuth 2.0 Support with the OIDC Client filter extension

OAuth 2.0 token management is handled by a Quarkus `OidcClient`. To enable this integration, you must add the Quarkus OIDC Client Filter extension to your project as shown in the following examples:

.Example of adding extension using Maven
[source,xml]
----
<dependency>
  <groupId>io.quarkus</groupId>

>   <artifactId>quarkus-oidc-client-filter</artifactId>

Reviewer: @gmunozfe @wmedvede I think we need to reassess this doc with https://docs.quarkiverse.io/quarkus-openapi-generator/dev/client.html#_oauth2_authentication

I'm unsure which version of the openapi-generator we are using on 1.36.

</dependency>

> ----

Reviewer 2: ```suggestion

<dependency>
  <groupId>io.quarkiverse.openapi.generator</groupId>
  <artifactId>quarkus-openapi-generator-oidc</artifactId>
  <version>2.9.0-lts</version>
</dependency>

----
```

Reviewer 2: Hi, @kaldesai in the sugestion above :point_up: we are adding basically one new dependency.
with the version -> **2.9.0-tls** 

Pelease take a look if instead of adding the fixed value **2.9.0-lts** we can define a document variable, and use it in this configuration. In other parts of the documentation  IIRC, we have already done this to setup for instance some image  tags numbers :thinking:

Author: Yes, I have added the attribute for the same

Reviewer 2: @kaldesai this change I don't see in the preview link https://92245--ocpdocs-pr.netlify.app/openshift-serverless/latest/serverless-logic/serverless-logic-managing-security/serverless-logic-authentication-openapi-services#serverless-logic-security-example-oauth-authentication_serverless-logic-authentication-openapi-services  :thinking: , maybe outdated?

Reviewer 2: Not sure if was didn't explain well of the merged wasn't good, let me explain:

I the produced document we must see two maven dependencies:

<dependency>
  <groupId>io.quarkus</groupId>
  <artifactId>quarkus-oidc-client-filter</artifactId>
</dependency>

<dependency>
  <groupId>io.quarkiverse.openapi.generator</groupId>
  <artifactId>quarkus-openapi-generator-oidc</artifactId>
  <version>2.9.0-lts</version>
</dependency>

But I in the preview link I can see this:

![image](https://github.com/user-attachments/assets/29e58501-8a74-4fe9-bb2e-74593d567be8)

Author: @wmedvede so there should be two dependencies, correct? As follows?  

[source,xml]
----
<dependency>
<groupId>io.quarkus</groupId>
  <artifactId>quarkus-oidc-client-filter</artifactId>
</dependency>

<dependency>
  <groupId>io.quarkiverse.openapi.generator</groupId>
  <artifactId>quarkus-openapi-generator-oidc</artifactId>
  <version>{ServerlessLogicOauthDependencyVersion}</version>
</dependency>
----

.Example of adding extension using Quarkus CLI
[source,terminal]
----
$ quarkus extension add quarkus-oidc-client-filter
----

[id="serverless-logic-security-oidc-configuration_{context}"]
== `OidcClient` configuration

To access the secured operation, define an `OidcClient` configuration in your `application.properties` file. The configuration uses the sanitized security scheme name from the OpenAPI specification, in this case, `oauth_example` as follows:

[source,terminal]
----
quarkus.oidc-client.oauth_example.auth-server-url=https://example.com/oauth
quarkus.oidc-client.oauth_example.token-path=/tokens
quarkus.oidc-client.oauth_example.discovery-enabled=false
quarkus.oidc-client.oauth_example.client-id=kogito-app
quarkus.oidc-client.oauth_example.grant.type=client
quarkus.oidc-client.oauth_example.credentials.client-secret.method=basic
quarkus.oidc-client.oauth_example.credentials.client-secret.value=secret
----

In this configuration:

> * `oauth_example` matches the sanitized name of the `oauth-example` scheme in the OpenAPI file.

Reviewer 2: ```suggestion
* `oauth_example` matches the sanitized name of the `oauth-example` scheme in the OpenAPI file. Finally, the link between the sanitized scheme name and the corresponding `OidcClient` is by using that simple naming convention.
```

// Module included in the following assemblies:
// serverles-logic/serverless-logic-authentication-openapi-services

:_mod-docs-content-type: REFERENCE
[id="serverless-logic-security-example-oauth-authentication_{context}"]

> = Example of OAuth 2.0 authentication

Author: Fixed

            }
          }
        },
        "security": [
          {
            "oauth-example": []
          }
        ]
      }
    }
  },
  "components": {
    "securitySchemes": {
      "oauth-example": {
        "type": "oauth2",
        "flows": {
          "clientCredentials": {
            "authorizationUrl": "https://example.com/oauth",
            "tokenUrl": "https://example.com/oauth/token",
            "scopes": {}
          }
        }
      }
    }
  }
}
----

In this example, the `sayHelloOauth2` operation is protected by the `oauth-example` security scheme, which uses the `clientCredentials` flow for token-based authentication.

[id="serverless-logic-security-oauth-support-oidc-client-filter-extention_{context}"]
== OAuth 2.0 Support with the OIDC Client filter extension

OAuth 2.0 token management is handled by a Quarkus `OidcClient`. To enable this integration, you must add the Quarkus OIDC Client Filter extension to your project as shown in the following examples:

.Example of adding extension using Maven
[source,xml]
----
<dependency>
  <groupId>io.quarkiverse.openapi.generator</groupId>
  <artifactId>quarkus-openapi-generator-oidc</artifactId>

>   <version>{ServerlessLogicOauthDependencyVerison}</version>

Reviewer 2: @kaldesai Did you try by using e.g. [source,xml,subs="attributes+"], with `+`  at the end of attributes+

                  "type": "string"
                }
              }
            }
          }
        },
        "security": [
          {
            "oauth-example": []
          }
        ]
      }
    }
  },
  "components": {
    "securitySchemes": {
      "oauth-example": {
        "type": "oauth2",
        "flows": {
          "clientCredentials": {
            "authorizationUrl": "https://example.com/oauth",
            "tokenUrl": "https://example.com/oauth/token",
            "scopes": {}
          }
        }
      }
    }
  }
}
----

In this example, the `sayHelloOauth2` operation is protected by the `oauth-example` security scheme, which uses the `clientCredentials` flow for token-based authentication.

[id="serverless-logic-security-oauth-support-oidc-client-filter-extention_{context}"]
== OAuth 2.0 Support with the OIDC Client filter extension

OAuth 2.0 token management is handled by a Quarkus `OidcClient`. To enable this integration, you must add the Quarkus OIDC Client Filter extension to your project as shown in the following examples:

.Example of adding extension using Maven
[source,xml]
----
<dependency>
<groupId>io.quarkus</groupId>

>   <artifactId>quarkus-oidc-client-filter</artifactId>

Reviewer 2: see my comment, I think using [source,xml,subs="attributes+"] could work to use variables substitution

</dependency>

<dependency>
  <groupId>io.quarkiverse.openapi.generator</groupId>
  <artifactId>quarkus-openapi-generator-oidc</artifactId>
  <version>2.9.0-lts</version>
</dependency>
----

> .Example of adding extension using Quarkus CLI

Author: Okay

> .Example of adding extension using Quarkus CLI

Author: Removed

[source,terminal]
----
$ quarkus extension add quarkus-oidc-client-filter
----

> 

Reviewer 2: See my comments, I think using [source,terminal,subs="attributes+"]  could work for using variables substitution

</dependency>

<dependency>
  <groupId>io.quarkiverse.openapi.generator</groupId>
  <artifactId>quarkus-openapi-generator-oidc</artifactId>
  <version>{ServerlessLogicOauthDependencyVersion}</version>
</dependency>
----

.Example of adding extension using `gitops` profile

Ensure that you configure the QUARKUS_EXTENSIONS build argument with the following value when building the workflow image:
[source,text,subs="attributes+"]
----
$ --build-arg=QUARKUS_EXTENSIONS=io.quarkus:quarkus-oidc-client-filter:{ServerlessLogicQuarkusVersion},io.quarkiverse.openapi.generator:quarkus-openapi-generator-oidc:{ServerlessLogicOauthDependencyVersion}
----

[id="serverless-logic-security-oidc-configuration_{context}"]
== `OidcClient` configuration

To access the secured operation, define an `OidcClient` configuration in your `application.properties` file. The configuration uses the sanitized security scheme name from the OpenAPI specification, in this case, `oauth_example` as follows:

[source,terminal]
----
quarkus.oidc-client.oauth_example.auth-server-url=https://example.com/oauth
quarkus.oidc-client.oauth_example.token-path=/tokens
quarkus.oidc-client.oauth_example.discovery-enabled=false

> quarkus.oidc-client.oauth_example.client-id=kogito-app

Reviewer 2: ```suggestion
quarkus.oidc-client.oauth_example.client-id=example-app
```

quarkus.oidc-client.oauth_example.grant.type=client
quarkus.oidc-client.oauth_example.credentials.client-secret.method=basic
quarkus.oidc-client.oauth_example.credentials.client-secret.value=secret
----

In this configuration:

* `oauth_example` matches the sanitized name of the `oauth-example` scheme in the OpenAPI file. The link between the sanitized scheme name and the corresponding `OidcClient` is by using that simple naming convention.

> * The OidcClient handles token generation and renewal automatically during workflow execution.

Reviewer 2: ```suggestion
* The OidcClient handles token generation and renewal automatically during workflow execution.

[NOTE]
====
For more information on these configurations you must refer to the Quakrus OpenID Connect and OAuth 2.0 documentation. 
====
```

Author: I think we can not mention the upstream docs like Quarkus OpenID Connect and OAuth 2.0 documentation in the prod docs

// Module included in the following assemblies:
// serverles-logic/serverless-logic-authentication-openapi-services

:_mod-docs-content-type: REFERENCE
[id="serverless-logic-security-example-oauth-authentication_{context}"]
= Example of clientCredentials OAuth 2.0 authentication

> The following example shows how to secure an OpenAPI operation using the OAuth 2.0 `clientCredentials` flow. The OpenAPI specification defines the `sayHelloOauth2` operation, which uses the `oauth-example` security scheme. Unlike simpler authentication methods such as HTTP Basic or API keys, OAuth 2.0 authentication requires additional integration with the Quarkus OpenID Connect (OIDC) Client.

Reviewer 3: ```suggestion
The following example shows how to secure an OpenAPI operation using the OAuth 2.0 `clientCredentials` flow. The OpenAPI specification defines the `sayHelloOauth2` operation, which uses the `oauth-example` security scheme. Unlike simpler authentication methods, such as HTTP Basic or API keys, OAuth 2.0 authentication requires additional integration with the Quarkus OpenID Connect (OIDC) Client.
```

    "title": "Oauth2 Scheme Example",
    "version": "1.0"
  },
  "paths": {
    "/hello-with-oauth2": {
      "get": {
        "operationId": "sayHelloOauth2",
        "responses": {
          "200": {
            "description": "OK",
            "content": {
              "text/plain": {
                "schema": {
                  "type": "string"
                }
              }
            }
          }
        },
        "security": [
          {
            "oauth-example": []
          }
        ]
      }
    }
  },
  "components": {
    "securitySchemes": {
      "oauth-example": {
        "type": "oauth2",
        "flows": {
          "clientCredentials": {
            "authorizationUrl": "https://example.com/oauth",
            "tokenUrl": "https://example.com/oauth/token",
            "scopes": {}
          }
        }
      }
    }
  }
}
----

In this example, the `sayHelloOauth2` operation is protected by the `oauth-example` security scheme, which uses the `clientCredentials` flow for token-based authentication.

[id="serverless-logic-security-oauth-support-oidc-client-filter-extention_{context}"]
== OAuth 2.0 Support with the OIDC Client filter extension

> OAuth 2.0 token management is handled by a Quarkus `OidcClient`. To enable this integration, you must add the Quarkus OIDC Client Filter extension to your project as shown in the following examples:

Reviewer 2: ```suggestion
OAuth 2.0 token management is handled by a Quarkus `OidcClient`. To enable this integration, you must add the Quarkus OIDC Client Filter, and the Quarkus OpenApi Generator OIDC extensions to your project as shown in the following examples:
```

> .Example of adding extension using Maven

Reviewer 2: ```suggestion
.Example of adding extensions using Maven
```

[source,text,subs="attributes+"]
----
<dependency>
  <groupId>io.quarkus</groupId>
  <artifactId>quarkus-oidc-client-filter</artifactId>
  <version>{ServerlessLogicQuarkusVersion}</version>
</dependency>

<dependency>
  <groupId>io.quarkiverse.openapi.generator</groupId>
  <artifactId>quarkus-openapi-generator-oidc</artifactId>
  <version>{ServerlessLogicOauthDependencyVersion}</version>
</dependency>
----

> .Example of adding extension using `gitops` profile

Reviewer 2: ```suggestion
.Example of adding extensions using `gitops` profile
```

Ensure that you configure the QUARKUS_EXTENSIONS build argument with the following value when building the workflow image:
[source,text,subs="attributes+"]
----

> $ --build-arg=QUARKUS_EXTENSIONS=io.quarkus:quarkus-oidc-client-filter:{ServerlessLogicQuarkusVersion},io.quarkiverse.openapi.generator:quarkus-openapi-generator-oidc:{ServerlessLogicOauthDependencyVersion}

Reviewer 2: ```suggestion
 --build-arg=QUARKUS_EXTENSIONS=io.quarkus:quarkus-oidc-client-filter:{ServerlessLogicQuarkusVersion},io.quarkiverse.openapi.generator:quarkus-openapi-generator-oidc:{ServerlessLogicOauthDependencyVersion}
```

Author: Per our Slack discussion, I am keeping the `$` sign here as it is a command.

----

> 

Reviewer 2: ```suggestion
.Example of adding extensions using `preview` profile

[source,yaml,subs="attributes+"]
----
apiVersion: sonataflow.org/v1alpha08
kind: SonataFlowPlatform
metadata:
  name: sonataflow-platform-example
  namespace: example-namespace
spec:
  build:
    template:
      buildArgs:
        - name: QUARKUS_EXTENSIONS
          value: io.quarkus:quarkus-oidc-client-filter:{ServerlessLogicQuarkusVersion},io.quarkiverse.openapi.generator:quarkus-openapi-generator-oidc:{ServerlessLogicOauthDependencyVersion}
----

[NOTE]
====
Extensions added in the `SonataFlowPlatform` CR are included for all the workflows that you deploy in that namespace with the `preview` profile.
====
```

[id="serverless-logic-security-oidc-configuration_{context}"]
== `OidcClient` configuration

To access the secured operation, define an `OidcClient` configuration in your `application.properties` file. The configuration uses the sanitized security scheme name from the OpenAPI specification, in this case, `oauth_example` as follows:

[source,terminal]
----
# adjust these configurations according with the authentication service.
quarkus.oidc-client.oauth_example.auth-server-url=https://example.com/oauth

> quarkus.oidc-client.oauth_example.token-path=/tokens

Reviewer 2: ```suggestion
quarkus.oidc-client.oauth_example.token-path=/token
```

quarkus.oidc-client.oauth_example.discovery-enabled=false
quarkus.oidc-client.oauth_example.client-id=example-app
quarkus.oidc-client.oauth_example.grant.type=client
quarkus.oidc-client.oauth_example.credentials.client-secret.method=basic
quarkus.oidc-client.oauth_example.credentials.client-secret.value=secret
----

In this configuration:

> * `oauth_example` matches the sanitized name of the `oauth-example` scheme in the OpenAPI file. The link between the sanitized scheme name and the corresponding `OidcClient` is by using that simple naming convention.

Reviewer 3: ```suggestion
* `oauth_example` matches the sanitized name of the `oauth-example` scheme in the OpenAPI file. The link between the sanitized scheme name and the corresponding `OidcClient` is achieved by using that simple naming convention.
```

Reviewer 3: Not sure if this is the right fix. Feel free to amend as needed.
